src/NCPkgFilterService.cc

```diff
@@ -0,0 +1,322 @@
+//
+//
+//	METHOD NAME : NCPkgServiceTag::NCPkgServiceTag
+//	METHOD TYPE : Constructor
+//
+//	DESCRIPTION :
+//
+
+NCPkgServiceTag::NCPkgServiceTag ( ZyppService servicePtr)
+    : YTableCell(std::string(" "))
+    , service (servicePtr)
+{
+
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::NCPkgServiceTable
+//	METHOD TYPE : Constructor
+//
+//	DESCRIPTION :
+//
+
+NCPkgServiceTable::NCPkgServiceTable( YWidget *parent, YTableHeader *tableHeader, NCPackageSelector *pkg )
+    :NCTable( parent, tableHeader )
+    ,packager(pkg)
+    ,repo_manager(new zypp::RepoManager())
+{
+   fillHeader();
+   fillServiceList();
+}
+
+bool NCPkgServiceTable::any_service()
+{
+    bool ret = std::any_of(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [](const zypp::Repository& repo) {
+        // if the repository does not belong to any service then the service name is empty
+        return !repo.info().service().empty();
+    });
+
+    yuiMilestone() << "Found a libzypp service: " << ret << std::endl;
+    return ret;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::fillHeader
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Fill header of servicesitories table (name + URL)
+//
+
+void NCPkgServiceTable::fillHeader()
+{
+    std::vector <std::string> header;
+
+    header.reserve(2);
+    header.push_back( "L" );
+    header.push_back( "L" + NCPkgStrings::PkgName() );
+
+    setHeader( header);
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::addLine
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Add one line (with tag) to the service table
+//
+
+void NCPkgServiceTable::addLine ( ZyppService r, const std::vector <std::string> & cols )
+{
+    //use default ctor, add cell in the next step
+    YTableItem *tabItem = new YTableItem();
+
+    //place tag (with service reference) to the 0th column
+    tabItem->addCell( new NCPkgServiceTag ( r ) );
+
+    // and append the rest (name, URL and stuff)
+    for ( unsigned i = 1; i < cols.size() + 1; ++i ) {
+	tabItem->addCell( cols[ i-1 ]);
+    }
+
+    // this is NCTable::addItem( tabItem );
+    //it actually appends the line to the table
+    addItem( tabItem );
+
+
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getTag
+//	METHOD TYPE : NCPkgServiceTag *
+//
+//	DESCRIPTION : Get tag of service table line on current index,
+//		      ( contains service reference)
+//
+
+NCPkgServiceTag* NCPkgServiceTable::getTag (const int & index )
+{
+   NCTableLine *line = myPad()->ModifyLine( index );
+   if ( !line )
+   {
+	return 0;
+   }
+
+   YTableItem *it = dynamic_cast<YTableItem*> (line->origItem() );
+
+   //get actual service tag from 0th column of the table
+   YTableCell *tcell = it->cell(0);
```

Comment:
It should never happen, of course, since you only put YTableItems into that thing, but in principle, a dynamic_cast can always fail and return 0, and then you'd dereference a null pointer here, resulting in a segfault, an unhappy user and an even more unhappy Y2Maintainer getting a weird bug report. Please add an _if_.

Reply:
OK, again a closer look results in  a *WTF*, because the dynamic cast is not necessary at all (for this receiver).